docs/builder/ref.md

```diff
@@ -0,0 +1,386 @@
+my_component.output("my_output", 32)
+```
+
+To access the input and output ports of a component within the definition of a component, use the syntax `my_component.this().port`.
+
+```python
+def add_my_component(prog):
+    my_component = prog.component("my_component")
+    my_component.output("my_output", 32)
+
+    with my_component.group("my_group"):
+        my_component.this().my_output = const(32, 1)
+```
+
+Note that it's possible to [create a handle][hndl] to input and output ports.
+
+### Multi-Component Designs
+
+Calyx supports [multi-component designs][multi]. The [top-level example][top] demonstrates how to construct multi-component designs using the library.
+
+#### Defining Common Calyx Cells
+
+Here's a snippet of code that adds a few common kinds of cells to a component:
+
+```python
+my_component = prog.component("my_component")
+
+# Registers: reg(name, bitwidth)
+my_component.reg("my_reg", 32)
+
+# Constants: const(name, bitwidth, value)
+my_component.const("my_reg", 32, 42)
+
+# Adders/Subtractors: [add|sub](name, size, signed=False)
+# a signed adder 
+my_component.add("my_add", 32, signed=True)
+# a subtractor
+my_component.sub("my_sub", 32)
+
+
+# Comparators: [gt|lt|eq|neq|ge|le](name, size, signed=False)
+my_component.gt("my_gt", 32)
+# a signed le comparison
+my_component.lt("my_lt", 32, signed=True)
+my_component.eq("my_eq", 32)
+my_component.neq("my_neq", 32)
+my_component.ge("my_ge", 32)
+my_component.le("my_le", 32)
+
+# 1-D memory: 
+# mem_d1(name, bitwidth, len, idx_size, is_external=False, is_ref=False)
+my_component.mem_d1("my_mem", 32, 4, 32)
+# An external memory
+my_component.mem_d1("my_mem", 32, 4, 32, is_external=True)
+# A memory by reference
+my_component.mem_d1("my_mem", 32, 4, 32, is_ref=True)
+```
+
+If you're curious, you can read more about [external memories][ext] or [memories by reference][ref].
+
+#### Retrieving Cells
+
+In order to reference a cell without a [handle][hndl], use the `Builder().get_cell()` method.
+
+```python
+# defining a register cell
+my_component.reg("my_reg", 32)
+
+# a few lines later 
+my_reg = prog.get_cell("my_reg")
+```
+
+## Wires
+
+### Guarded Assignments
+
+Guarded assignments in the builder are syntactically similar to those in Calyx.
+
+```python
+my_component = prog.component("my_component")
+
+my_add = comp.add("my_add", 32)
+my_reg = comp.reg("my_reg", 32)
+
+with my_component.group("my_group"):
+    # unconditional assignments
+    add.left = const(32, 1)
+    add.right = const(32, 41)
+    my_reg.write_en = 1
+
+    # a guarded assignment using @
+    my_reg.in = (add.left < add.right) @ add.out
+```
+
+### Groups
+
+#### Defining Groups
+
+[Groups][grp] are defined using the `group()` method for a component.  It's also possible to define a [static delay][static] for a group using the optional `static_delay` argument.
+
+```python
+my_component = prog.component("my_component")
+
+# a group definition
+with my_component.group("my_group"):
```

Comment:
we want to mention that the standard way to get a handle for a group (which is required to set done ports on non-combinational groups) is with the `as` syntax in python

```python
with my_component.group("my_group") as my_group:
    # assignments
    my_group.done = #stuff

```

Reply:
This probably merits an explicit mention, even though you show some of this syntax in the big block up top.